File: agent-go/internal/tools/grep.go
package tools

import (
	"bufio"
	"context"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

const grepMaxResults = 1000

type GrepTool struct{ cwd string }

func NewGrepTool(cwd string) *GrepTool { return &GrepTool{cwd: cwd} }

func (t *GrepTool) Name() string        { return "grep" }
func (t *GrepTool) Description() string { return "Search for a regex pattern in files. Returns file:line:content matches." }
func (t *GrepTool) Schema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path":        map[string]any{"type": "string", "description": "File or directory to search"},
			"pattern":     map[string]any{"type": "string", "description": "Regular expression pattern"},
			"recursive":   map[string]any{"type": "boolean", "description": "Recurse into subdirectories (default: true)"},
			"ignore_case": map[string]any{"type": "boolean", "description": "Case-insensitive search (default: false)"},
		},
		"required": []string{"path", "pattern"},
	}
}

func (t *GrepTool) Execute(_ context.Context, _ string, args map[string]any, _ UpdateFunc) (ToolResult, error) {
	root, _ := args["path"].(string)
	pattern, _ := args["pattern"].(string)
	if root == "" {
		return errResult("grep: missing required parameter 'path'"), nil
	}
	if pattern == "" {
		return errResult("grep: missing required parameter 'pattern'"), nil
	}
	root = resolvePath(root, t.cwd)

	ignoreCase := false
	if v, ok := args["ignore_case"].(bool); ok {
		ignoreCase = v
	}
	recursive := true
	if v, ok := args["recursive"].(bool); ok {
		recursive = v
	}

	reStr := pattern
	if ignoreCase {
		reStr = "(?i)" + pattern
	}
	re, err := regexp.Compile(reStr)
	if err != nil {
		return errResult(fmt.Sprintf("grep: invalid pattern: %v", err)), nil
	}

	var results []string
	walkFn := func(path string, d fs.DirEntry, err error) error {
		if err != nil || d.IsDir() {
			if d != nil && d.IsDir() && !recursive && path != root {
				return fs.SkipDir
			}
			return nil
		}
		f, err := os.Open(path)
		if err != nil {
			return nil
		}
		defer f.Close()

		rel, _ := filepath.Rel(t.cwd, path)
		scanner := bufio.NewScanner(f)
		lineNo := 0
		for scanner.Scan() {
			lineNo++
			line := scanner.Text()
			if re.MatchString(line) {
				results = append(results, fmt.Sprintf("%s:%d:%s", rel, lineNo, line))
				if len(results) >= grepMaxResults {
					return fmt.Errorf("limit reached")
				}
			}
		}
		return nil
	}

	fi, err := os.Stat(root)
	if err != nil {
		return errResult(fmt.Sprintf("grep: %v", err)), nil
	}
	if fi.IsDir() {
		_ = filepath.WalkDir(root, walkFn)
	} else {
		_ = walkFn(root, &fileEntry{fi}, nil)
	}

	if len(results) == 0 {
		return ToolResult{Content: "(no matches)"}, nil
	}
	out := strings.Join(results, "\n")
	if len(results) >= grepMaxResults {
		out += fmt.Sprintf("\n[Truncated at %d matches]", grepMaxResults)
	}
	return ToolResult{Content: out}, nil
}

// fileEntry wraps os.FileInfo to implement fs.DirEntry for single-file grep.
type fileEntry struct{ fi os.FileInfo }

func (e *fileEntry) Name() string               { return e.fi.Name() }
func (e *fileEntry) IsDir() bool                { return e.fi.IsDir() }
func (e *fileEntry) Type() fs.FileMode          { return e.fi.Mode().Type() }
func (e *fileEntry) Info() (fs.FileInfo, error) { return e.fi, nil }

File: agent-python/src/pi/tools/grep.py
# src/pi/tools/grep.py
from __future__ import annotations

import asyncio
import re
from pathlib import Path
from typing import Any, Callable

from .base import Tool, ToolResult


class GrepTool(Tool):
    @property
    def name(self) -> str:
        return "grep"

    @property
    def description(self) -> str:
        return "Search for a regex pattern in a file or directory. Returns file:line:content."

    @property
    def parameters(self) -> dict:
        return {
            "type": "object",
            "properties": {
                "path": {"type": "string"},
                "pattern": {"type": "string", "description": "Regex pattern"},
                "recursive": {"type": "boolean", "default": True},
                "ignore_case": {"type": "boolean", "default": False},
            },
            "required": ["path", "pattern"],
        }

    async def execute(
        self,
        tool_call_id: str,
        args: dict[str, Any],
        signal: asyncio.Event | None = None,
        on_update: Callable[[str], None] | None = None,
    ) -> ToolResult:
        path = Path(args.get("path", "."))
        pattern_str = args.get("pattern", "")
        recursive = args.get("recursive", True)
        ignore_case = args.get("ignore_case", False)

        flags = re.IGNORECASE if ignore_case else 0
        try:
            regex = re.compile(pattern_str, flags)
        except re.error as e:
            return ToolResult.err(f"Invalid regex pattern: {e}")

        files: list[Path] = []
        if path.is_file():
            files = [path]
        elif path.is_dir():
            files = sorted(path.rglob("*") if recursive else path.iterdir())
            files = [f for f in files if f.is_file()]
        else:
            return ToolResult.err(f"Path not found: {path}")

        results: list[str] = []
        for file in files:
            try:
                text = file.read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue
            for lineno, line in enumerate(text.splitlines(), 1):
                if regex.search(line):
                    results.append(f"{file}:{lineno}:{line}")

        if not results:
            return ToolResult.ok("No matches found.")
        return ToolResult.ok("\n".join(results))
